api/src/main/java/io/minio/MinioClient.java

```diff
@@ -969,6 +969,9 @@ private Request createRequest(Method method, String bucketName, String objectNam
         } else {
           // Fix issue #567: Compute SHA256 hash only.
           sha256Hash = Digest.sha256Hash(data, len);
+          if ( method == Method.HEAD ) {
+            requestBuilder.header("Accept-Encoding", "identity");
```

Comment:
I don't know disabling transparent gzip on response bodies is correct here.  I would like to understand more. Please provide
1. http traces by using [traceOn()](http://minio.github.io/minio-java/io/minio/MinioClient.html#traceOn-java.io.OutputStream-)
2. http traces of `minio-go` or `mc`

Reply:
I've logs from nginx & mc. I'm looking at it to extract those logs. I am also writing another solution that modified only the response.

Reply:
I cannot attach the file with debug extracts here because of some errors. I tried adding an interceptor for gzip content. It didn't work for all the variations I tried. If we compare it with minio-py & minio-go, the results come down to Accept-encoding type. minio-py always seems to send sends Accept-encoding as identity for the various statobject requests I have tried. From various comments I've read by the developers they set Accept-Encoding to only identity or chunked, but this might change in the future. In debug logs however I see that Accept-encoding is not set, which is probably why we don't get a gzip response.

Reply:
```
priyankminio1.com 
priyankminio1.com 
priyankminio1.com [REQUEST s3.GetBucketLocation] 23:12:28.870
priyankminio1.com GET /photos?location=
priyankminio1.com Host: priyankminio1.com
priyankminio1.com X-Forwarded-Proto: http
priyankminio1.com X-Real-Ip: 192.168.1.8
priyankminio1.com Content-Length: 0
priyankminio1.com X-Amz-Date: 20191206T231228Z
priyankminio1.com X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
priyankminio1.com X-Forwarded-For: 192.168.1.8
priyankminio1.com Accept-Encoding: gzip
priyankminio1.com Authorization: AWS4-HMAC-SHA256 Credential=minio/20191206/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=0c644df63e1e3273827868546c0f6dd3b6113829c8b722669f7df61cc88eecaa
priyankminio1.com User-Agent: MinIO (amd64; amd64) minio-java/dev
priyankminio1.com 
priyankminio1.com [RESPONSE] [23:12:28.870] [ Duration 160µs  ↑ 137 B  ↓ 367 B ]
priyankminio1.com 200 OK
priyankminio1.com Vary: Origin
priyankminio1.com X-Amz-Request-Id: 15DDEB1BB443DAE5
priyankminio1.com X-Xss-Protection: 1; mode=block
priyankminio1.com Accept-Ranges: bytes
priyankminio1.com Content-Length: 128
priyankminio1.com Content-Security-Policy: block-all-mixed-content
priyankminio1.com Content-Type: application/xml
priyankminio1.com Server: MinIO/DEVELOPMENT.GOGET
priyankminio1.com <?xml version="1.0" encoding="UTF-8"?>
<LocationConstraint xmlns="http://s3.amazonaws.com/doc/2006-03-01/"></LocationConstraint>
priyankminio1.com 
priyankminio1.com [REQUEST s3.HeadObject] 23:12:28.899
priyankminio1.com HEAD /photos/priyraj-cisco.jpg
priyankminio1.com Host: priyankminio1.com
priyankminio1.com Accept-Encoding: gzip
priyankminio1.com User-Agent: MinIO (amd64; amd64) minio-java/dev
priyankminio1.com X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
priyankminio1.com X-Amz-Date: 20191206T231228Z
priyankminio1.com X-Forwarded-For: 192.168.1.8
priyankminio1.com X-Forwarded-Proto: http
priyankminio1.com Authorization: AWS4-HMAC-SHA256 Credential=minio/20191206/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=fe6a407883a094fd6dfb3dcf6e526739b9febb663e66840d5686ba03e66259fe
priyankminio1.com Content-Length: 0
priyankminio1.com X-Real-Ip: 192.168.1.8
priyankminio1.com 
priyankminio1.com [RESPONSE] [23:12:28.899] [ Duration 203µs  ↑ 137 B  ↓ 324 B ]
priyankminio1.com 200 OK
priyankminio1.com Content-Type: image/jpeg
priyankminio1.com ETag: "00000000000000000000000000000000-1"
priyankminio1.com Vary: Origin
priyankminio1.com X-Xss-Protection: 1; mode=block
priyankminio1.com Accept-Ranges: bytes
priyankminio1.com Content-Length: 14780
priyankminio1.com Content-Security-Policy: block-all-mixed-content
priyankminio1.com Last-Modified: Wed, 04 Dec 2019 19:27:31 GMT
priyankminio1.com Server: MinIO/DEVELOPMENT.GOGET
priyankminio1.com X-Amz-Request-Id: 15DDEB1BB5F9B7DA

priyank@priyank-MacBookPro: examples$ java -cp "/home/priyank/ws/minio/src/github/minio-java/api/build/libs/minio-6.0.12-DEV-all.jar:." StatObject
ObjectStat{bucket='photos', name='priyraj-cisco.jpg', contentType='image/jpeg', createdTime=Wed Dec 04 11:27:31 PST 2019, length=0, etag='W/00000000000000000000000000000000-1'}

2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream process non buffered upstream
2019/12/06 11:12:28 [debug] 1276#1276: *170 recv: eof:0, avail:1
2019/12/06 11:12:28 [debug] 1276#1276: *170 recv: fd:17 6 of 4096
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 31 s:0
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0D s:1
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0A s:3
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 20 s:4
2019/12/06 11:12:28 [debug] 1276#1276: *170 http proxy out buf 000055820A9DF3C3 1
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0D s:5
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0A s:6
2019/12/06 11:12:28 [debug] 1276#1276: *170 http output filter "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http copy filter: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http postpone filter "/minio/admin/v2/trace?all=true&err=false" 000055820A9EC2B0
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunk: 1
2019/12/06 11:12:28 [debug] 1276#1276: *170 write new buf t:1 f:0 000055820A9E05A0, pos 000055820A9E05A0, size: 3 file: 0, size: 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 write new buf t:0 f:0 0000000000000000, pos 000055820A9DF3C3, size: 1 file: 0, size: 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 write new buf t:0 f:0 000055820A9E0508, pos 0000558209465C7A, size: 2 file: 0, size: 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 http write filter: l:0 f:1 s:6
2019/12/06 11:12:28 [debug] 1276#1276: *170 http write filter limit 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 writev: 6 of 6
2019/12/06 11:12:28 [debug] 1276#1276: *170 http write filter 0000000000000000
2019/12/06 11:12:28 [debug] 1276#1276: *170 http copy filter: 0 "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 event timer del: 17: 7830794
2019/12/06 11:12:28 [debug] 1276#1276: *170 event timer add: 17: 60000:7831294
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream dummy handler
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream process non buffered upstream
2019/12/06 11:12:28 [debug] 1276#1276: *170 recv: eof:0, avail:1
2019/12/06 11:12:28 [debug] 1276#1276: *170 recv: fd:17 6 of 4096
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 31 s:0
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0D s:1
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0A s:3
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 20 s:4
2019/12/06 11:12:28 [debug] 1276#1276: *170 http proxy out buf 000055820A9DF3C3 1
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0D s:5
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunked byte: 0A s:6
2019/12/06 11:12:28 [debug] 1276#1276: *170 http output filter "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http copy filter: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http postpone filter "/minio/admin/v2/trace?all=true&err=false" 000055820A9EC2B0
2019/12/06 11:12:28 [debug] 1276#1276: *170 http chunk: 1
2019/12/06 11:12:28 [debug] 1276#1276: *170 write new buf t:1 f:0 000055820A9E0508, pos 000055820A9E0508, size: 3 file: 0, size: 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 write new buf t:0 f:0 0000000000000000, pos 000055820A9DF3C3, size: 1 file: 0, size: 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 write new buf t:0 f:0 000055820A9E05A0, pos 0000558209465C7A, size: 2 file: 0, size: 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 http write filter: l:0 f:1 s:6
2019/12/06 11:12:28 [debug] 1276#1276: *170 http write filter limit 0
2019/12/06 11:12:28 [debug] 1276#1276: *170 writev: 6 of 6
2019/12/06 11:12:28 [debug] 1276#1276: *170 http write filter 0000000000000000
2019/12/06 11:12:28 [debug] 1276#1276: *170 http copy filter: 0 "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 event timer del: 17: 7831294
2019/12/06 11:12:28 [debug] 1276#1276: *170 event timer add: 17: 60000:7831794
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:12:28 [debug] 1276#1276: *170 http upstream dummy handler




priyankminio1.com 
priyankminio1.com [RESPONSE] [23:14:17.004] [ Duration 144µs  ↑ 137 B  ↓ 367 B ]
priyankminio1.com 200 OK
priyankminio1.com X-Xss-Protection: 1; mode=block
priyankminio1.com Accept-Ranges: bytes
priyankminio1.com Content-Length: 128
priyankminio1.com Content-Security-Policy: block-all-mixed-content
priyankminio1.com Content-Type: application/xml
priyankminio1.com Server: MinIO/DEVELOPMENT.GOGET
priyankminio1.com Vary: Origin
priyankminio1.com X-Amz-Request-Id: 15DDEB34E18C7EEC
priyankminio1.com <?xml version="1.0" encoding="UTF-8"?>
<LocationConstraint xmlns="http://s3.amazonaws.com/doc/2006-03-01/"></LocationConstraint>
priyankminio1.com 
priyankminio1.com [REQUEST s3.HeadObject] 23:14:17.033
priyankminio1.com HEAD /photos/priyraj-cisco.jpg
priyankminio1.com Host: priyankminio1.com
priyankminio1.com X-Forwarded-Proto: http
priyankminio1.com X-Real-Ip: 192.168.1.8
priyankminio1.com Authorization: AWS4-HMAC-SHA256 Credential=minio/20191206/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;host;x-amz-content-sha256;x-amz-date, Signature=c368c0705954434174147e2cb447b68889d372bd121abdb95d1afe86cda0c947
priyankminio1.com Content-Length: 0
priyankminio1.com X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
priyankminio1.com X-Amz-Date: 20191206T231417Z
priyankminio1.com Accept-Encoding: identity
priyankminio1.com User-Agent: MinIO (amd64; amd64) minio-java/dev
priyankminio1.com X-Forwarded-For: 192.168.1.8
priyankminio1.com 
priyankminio1.com [RESPONSE] [23:14:17.033] [ Duration 188µs  ↑ 137 B  ↓ 324 B ]
priyankminio1.com 200 OK
priyankminio1.com Content-Type: image/jpeg
priyankminio1.com Server: MinIO/DEVELOPMENT.GOGET
priyankminio1.com Vary: Origin
priyankminio1.com X-Amz-Request-Id: 15DDEB34E33F5BED
priyankminio1.com X-Xss-Protection: 1; mode=block
priyankminio1.com Accept-Ranges: bytes
priyankminio1.com Content-Security-Policy: block-all-mixed-content
priyankminio1.com Last-Modified: Wed, 04 Dec 2019 19:27:31 GMT
priyankminio1.com Content-Length: 14780
priyankminio1.com ETag: "00000000000000000000000000000000-1"

priyank@priyank-MacBookPro: examples$ java -cp "/home/priyank/ws/minio/src/github/minio-java/api/build/libs/minio-6.0.12-DEV-all.jar:." StatObject
ObjectStat{bucket='photos', name='priyraj-cisco.jpg', contentType='image/jpeg', createdTime=Wed Dec 04 11:27:31 PST 2019, length=14780, etag='00000000000000000000000000000000-1'}


2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream process non buffered upstream
2019/12/06 11:14:17 [debug] 1276#1276: *170 recv: eof:0, avail:1
2019/12/06 11:14:17 [debug] 1276#1276: *170 recv: fd:17 6 of 4096
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 31 s:0
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0D s:1
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0A s:3
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 20 s:4
2019/12/06 11:14:17 [debug] 1276#1276: *170 http proxy out buf 000055820A9DF3C3 1
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0D s:5
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0A s:6
2019/12/06 11:14:17 [debug] 1276#1276: *170 http output filter "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http copy filter: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http postpone filter "/minio/admin/v2/trace?all=true&err=false" 000055820A9EC2B0
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunk: 1
2019/12/06 11:14:17 [debug] 1276#1276: *170 write new buf t:1 f:0 000055820A9E05A0, pos 000055820A9E05A0, size: 3 file: 0, size: 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 write new buf t:0 f:0 0000000000000000, pos 000055820A9DF3C3, size: 1 file: 0, size: 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 write new buf t:0 f:0 000055820A9E0508, pos 0000558209465C7A, size: 2 file: 0, size: 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 http write filter: l:0 f:1 s:6
2019/12/06 11:14:17 [debug] 1276#1276: *170 http write filter limit 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 writev: 6 of 6
2019/12/06 11:14:17 [debug] 1276#1276: *170 http write filter 0000000000000000
2019/12/06 11:14:17 [debug] 1276#1276: *170 http copy filter: 0 "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 event timer del: 17: 7939794
2019/12/06 11:14:17 [debug] 1276#1276: *170 event timer add: 17: 60000:7940294
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream dummy handler
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream process non buffered upstream
2019/12/06 11:14:17 [debug] 1276#1276: *170 recv: eof:0, avail:1
2019/12/06 11:14:17 [debug] 1276#1276: *170 recv: fd:17 6 of 4096
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 31 s:0
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0D s:1
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0A s:3
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 20 s:4
2019/12/06 11:14:17 [debug] 1276#1276: *170 http proxy out buf 000055820A9DF3C3 1
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0D s:5
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunked byte: 0A s:6
2019/12/06 11:14:17 [debug] 1276#1276: *170 http output filter "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http copy filter: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http postpone filter "/minio/admin/v2/trace?all=true&err=false" 000055820A9EC2B0
2019/12/06 11:14:17 [debug] 1276#1276: *170 http chunk: 1
2019/12/06 11:14:17 [debug] 1276#1276: *170 write new buf t:1 f:0 000055820A9E0508, pos 000055820A9E0508, size: 3 file: 0, size: 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 write new buf t:0 f:0 0000000000000000, pos 000055820A9DF3C3, size: 1 file: 0, size: 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 write new buf t:0 f:0 000055820A9E05A0, pos 0000558209465C7A, size: 2 file: 0, size: 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 http write filter: l:0 f:1 s:6
2019/12/06 11:14:17 [debug] 1276#1276: *170 http write filter limit 0
2019/12/06 11:14:17 [debug] 1276#1276: *170 writev: 6 of 6
2019/12/06 11:14:17 [debug] 1276#1276: *170 http write filter 0000000000000000
2019/12/06 11:14:17 [debug] 1276#1276: *170 http copy filter: 0 "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 event timer del: 17: 7940294
2019/12/06 11:14:17 [debug] 1276#1276: *170 event timer add: 17: 60000:7940794
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream request: "/minio/admin/v2/trace?all=true&err=false"
2019/12/06 11:14:17 [debug] 1276#1276: *170 http upstream dummy handler

```

Reply:
By analysing log traces, I don't see any missing `Content-Length` is response. But I see you got `length=0`  

```bash
priyank@priyank-MacBookPro: examples$ java -cp "/home/priyank/ws/minio/src/github/minio-java/api/build/libs/minio-6.0.12-DEV-all.jar:." StatObject
ObjectStat{bucket='photos', name='priyraj-cisco.jpg', contentType='image/jpeg', createdTime=Wed Dec 04 11:27:31 PST 2019, length=0, etag='W/00000000000000000000000000000000-1'}
```

However there is no relevant log I see in your comment where a missing `Content-Length`  header in response.

Reply:
Yes. One of my previous comments is missing or wasn't submitted in this thread. _Content-Length_ is missing or 0 in minio-java/okhttp3 responses even though minio server & nginx are displaying the right values with the 'gzip on;' setting on nginx. I will push the other interceptor code for perusal.

Reply:
My suggestion is to reproduce the issue properly and provide logs using `traceOn()`. Any other method of collecting logs is not useful because `Response` object contains headers and body actually used for further processing.

Reply:
@balamurugana 
As explained in this document http://nginx.org/en/docs/http/ngx_http_gzip_module.html

```
The ngx_http_gzip_module module is a filter that compresses responses using the “gzip” method. This often helps to reduce the size of transmitted data by half or even more.
```
Nginx can be used to serve web services other than S3, which will make use of this feature.
This setting is a top level setting at Nginx, so it applies to all the web services that are being serviced by this Nginx.

As explained in thread https://forums.aws.amazon.com/thread.jspa?threadID=48276 and verified by testing, AWS S3 and MinIO ignore when the client sends `Accept-Encoding: gzip` header. But the Nginx that sits in the middle will try to compress data in gzip format. This is unexpected as far S3 clients are concerned. So, the correct solution to this issue is to set `Accept-Encoding: identity` which will force Nginx to not (de)compress the data flowing through it.

This issue can also be replicated using minio-go by setting `Acept-Encoding: gzip` header, so the correct solution to this issue is to make sure that all the SDKs that we have should set the `Accept-Encoding: identity` header.

Also, this is not an issue with aws-sdk-java as by default it does not set `Accept-Encoding` header. This is a problem with `minio-java` as we use `ok http` and that adds `Accept-Encoding: gzip` by default on all as explained in the ok http documentation

```
OkHttp may add headers that are absent from the original request, including `Content-Length`, `Transfer-Encoding`, `User-Agent`, `Host`, `Connection`, and `Content-Type`. It will add an `Accept-Encoding` header for transparent response compression unless the header is already present. If you’ve got cookies, OkHttp will add a `Cookie` header with them.
```